### ISSUE TITLE
Add static and dynamic batching support?

### ISSUE THREAD

**Issue by marynate** [godotengine/godot#1527]
I believe the performance of Godot can benefit from minimizing drawcalls via  static and dynamic batching support.

Here's what I found out from the [5000 sprites benchmark](https://github.com/marynate/godot.benchmark) :
- Testing 5000 sprites on screen without dynamic batching:

> Unity 5: 50001 drawcalls, 173 fps
> Godot: 50001 drawcalls, 175 fps

By turning off dynamic batching feature, Unity 5 and Godot has almost identical fps.
- Testing 5000 sprites on screen with dynamic batching:

> Unity 5: 2 drawcalls, 278 fps
> Godot: n/a
> Urho3D; ? drawcalls: 785 fps

WIth Dynamic batching, the drawcalls could be dramatically reduced hence the improvement of fps on Unity 5.

**Reference**:
[Dynamic and Static Batching in Unity](http://docs.unity3d.com/Manual/DrawCallBatching.html)
[OpenGL Batch Rendering](http://www.gamedev.net/page/reference/index.html/_/technical/opengl/opengl-batch-rendering-r3900)
[Sprite Batching in OpenGL ES 2.0](http://antonholmquist.com/blog/sprite-batching-opengl-es-2-0/)
[SpriteBatch in libgdx](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/graphics/g2d/SpriteBatch.java)
[CCSpriteBatchNode in cocos2d-x](https://github.com/cocos2d/cocos2d-x/blob/8716164a5d95480b7d61d114c215b4c4cb762077/cocos/2d/CCSpriteBatchNode.cpp)
[Batch in Urho3D](https://github.com/urho3d/Urho3D/blob/90833996fa797d6ecbac27ed78123b85fb974b42/Source/Urho3D/Graphics/Batch.cpp)

**Comment by reduz** [comment]
The drawing code in Godot has been rewritten recently to allow
optimizations such as this. Dynamic batching could be added now, but I
think there are other, more interestig, optimizations that can be attempted
first. This is why having benchmarks is vital :P

**Comment by marynate** [comment]
@reduz have you started working on your optimization magic? :)

**Comment by reduz** [comment]
Nope, 1.1 comes first.. Are there enough benchmarks for doing optimization?

**Comment by crookedbard** [comment]
any progress with this request?

**Comment by adeluiz** [comment]
The only thing that stops me from developing with Godot is this.

Looking for benchmarks (something most people do when looking for a new engine) I've found that Godot lacks of sprite batching (try the bunnymarks numbers for comparing engines: https://www.reddit.com/r/godot/comments/47j49k/gdscript_is_very_slow/). 

I hope that sprite batching could be implemented within the 3.0 version

**Comment by tagcup** [comment]
That benchmark only shows that GDScript is slow, it doesn't have much to do with rendering bottlenecks.

**Comment by Giacom** [comment]
The lack of sprite batching is quite limiting if you are looking to create low end mobile games.

Even an alternative, such as the multi-mesh node, would be useful. Built in dynamic batching would be best though for new godot developers.

**Comment by Demmios** [comment]
not only low end mobile games but low end pcs too i need this feature too my game cant handle a 200x200 tiles map wihout dropping the fps from 60 to 30 when zoom out
+1 for batching - fingers crossed

**Comment by Demmios** [comment]
As @eons suggested on discord i am ataching the project files with the Map Generator to help devs to see the issue and maybe point to a solution or the implementation of the batching
[Map Generator fps.zip](https://github.com/godotengine/godot/files/1616064/Map.Generator.fps.zip)

**Comment by eon-s** [comment]
I have tried the map generator and replacing the sprites with simple Node2D I get the same performance issues, trying to discard a debugger issue tested with Node too but these do not affect performance, only CanvasItems (with and without textures) seems to be the problem here.

**Comment by eon-s** [comment]
Well, tested with release builds (release and debug) and both run at 60fps with 40k sprites (double than editor run), something is being added by the editor and not just the number of nodes is what affects it.

**Comment by Demmios** [comment]
I have done some windows builds and running off the editor in my tests the fps are the same from editor or from the exported run =(

**Comment by ryancheung** [comment]
Maybe godot engine can add Sprite-Batch based drawing API for 2D, like what it can do in MonoGame or DirectX2D:
```charpt
protected override void Draw(GameTime gameTime)
{
graphics.GraphicsDevice.Clear(Color.CornflowerBlue);
 
// TODO: Add your drawing code here
spriteBatch.Begin();
spriteBatch.Draw(textureBall, new Vector2(0, 0), Color.White);
spriteBatch.End();
 
base.Draw(gameTime);
}
```
Code example are from http://www.monogame.net/documentation/?page=adding_content

It's also easy to add custom blending factors. 
When you have to dynamic draw tilemaps (e.g. create a game like LegendOfMir), it seems much easier to draw 2D the Sprite-Batch way than the node-based way.
![mir2](https://user-images.githubusercontent.com/240899/52908300-24867280-32ae-11e9-9a3d-ccc251a2657f.jpg)

**Comment by Zireael07** [comment]
For the above case, there's the Tilemap node.

But I would still like to know the status of dynamic batching?

**Comment by ryancheung** [comment]
Default tilemap node is not capable, since the map is not manually created in tilemap editor, it’s defined in custom map data, and so are the cell textures, the only to do is to create your own tilemap node with Godoy custom module.

The key point:
In node-based 2d drawing, blending or shader relys on canvas item nodes or canvas item created with VisualServer.
But in the LengendOfMir game, blending or shader must be able to changed on every texture draw-call, just the feature the common SpriteBatch API provided.
And with GodotEngine, there’s no full blend control.

So node-based drawing has great convenience on creating games in Game Editor, but it also lose convenience SpriteBatch-based provided.
It’s really hard to create a game like LegendOfMir with GodotEngone.

**Comment by reduz** [comment]
Such an old issue... I will be working on this for Godot 4.0 by the way for the GLES2 renderer. Vulkan will probably not need it given how it's going to be implemented.

**Comment by mrcdk** [comment]
@ryancheung if you change the shader, the blending function or the texture batching won't work because you need to modify the opengl state and you'll need to start a new batch per change. 

Batching is used because it creates 1 draw-call which will draw as many objects as you give it with one set of parameters. If you start modifying those parameters for every object you lose any benefit batching will give you.

**Comment by ryancheung** [comment]
@mrcdk It's OK to end previous batch and start new batch, since we don't need put all draw calls in one batch.

**Comment by yusdacra** [comment]
So, when this will be implemented for both GLES 3 and GLES 2? Even the 3D platformer demo (which is really simple) runs bad. This is probably the only thing preventing me from using Godot full time.

**Comment by reduz** [comment]
@yusdacra For 3D it will never be implemented, if you have bad performance in the 3D platformer demo it's probably related to something else.

Godot 4.0 will come with more advanced tools to measure better where your performance is being taken.

**Comment by fire** [comment]
For some background:

My opinion is that batching aka mesh merging should be done at content creation time or as a editor plugin, but not at runtime.

The merging operation takes multiple minutes for a large 3d scene and that's not feasible to do at runtime.

**Comment by red1939** [comment]
Probably covered by: #37349

**Comment by Calinou** [comment]
@red1939 That pull request only implements batching in 2D and when using the GLES2 renderer.

reduz has plans to further implement automatic *instancing* in the Vulkan renderer in both 2D and 3D so larger amounts of instances can be drawn.

**Comment by red1939** [comment]
Good to know! Will it be considerably different than sort & group by state
(texture, model, material, blending) + uber_shader.draw()?

**Comment by akien-mga** [comment]
Fixed by #37349. This will be ported to GLES3 too, and should be included in Godot 4.0 as well.
